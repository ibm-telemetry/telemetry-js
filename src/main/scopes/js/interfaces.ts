/*
 * Copyright IBM Corp. 2024, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import type * as ts from 'typescript'

import { Logger } from '../../core/log/logger.js'
import { ComplexValue } from './complex-value.js'
import { JsNodeHandler } from './node-handlers/js-node-handler.js'
import { NodeValueHandler } from './node-handlers/value-handlers/node-value-handler.js'

export interface JsImport {
  name: string
  path: string
  isDefault: boolean
  isAll: boolean
  rename?: string
}

type JsNodeHandlerClass<DataType> = new (
  node: ts.SourceFile,
  logger: Logger
) => JsNodeHandler<DataType>

export type JsNodeHandlerMap = Partial<Record<ts.SyntaxKind, JsNodeHandlerClass<unknown>>>

export type NodeValue = string | number | boolean | ComplexValue | null | undefined

type NodeValueHandlerProducer = new (node: ts.SourceFile, logger: Logger) => NodeValueHandler

export type NodeValueHandlerMap = Partial<Record<ts.SyntaxKind, NodeValueHandlerProducer>>
