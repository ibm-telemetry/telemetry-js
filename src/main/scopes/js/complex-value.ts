/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

/**
 * Object representing a complex value.
 */
export class ComplexValue {
  constructor(public readonly complexValue: unknown) {}
}
