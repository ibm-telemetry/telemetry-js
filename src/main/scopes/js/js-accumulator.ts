/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import type { JsImport } from '../js/interfaces.js'

/**
 * Base class for all JS Accumulators.
 * Responsible for maintaining an aggregated state of imports and other elements.
 */
export abstract class JsAccumulator {
  public imports: JsImport[]

  constructor() {
    this.imports = []
  }
}
