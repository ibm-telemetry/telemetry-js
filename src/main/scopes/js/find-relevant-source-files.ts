/*
 * Copyright IBM Corp. 2024, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import getPropertyByPath = require('lodash/get.js')
import * as path from 'node:path'

import { ObjectPath } from 'object-scan'

import { getTrackedSourceFiles } from '../../core/get-tracked-source-files.js'
import { Logger } from '../../core/log/logger.js'
import { NoInstallationFoundError } from '../../exceptions/no-installation-found-error.js'
import { getDependencyTree } from '../npm/get-dependency-tree.js'
import { getDirectoryPrefix } from '../npm/get-directory-prefix.js'
import { getInstalledVersionPaths } from '../npm/get-installed-version-paths.js'
import { getPackageData } from '../npm/get-package-data.js'
import { getPackageTrees } from '../npm/get-package-trees.js'
import { getTreePredecessor } from '../npm/get-tree-predecessor.js'
import { DependencyTree, PackageData } from '../npm/interfaces.js'

/**
 * Finds tracked source files and then filters them based on ones that appear in a  project which
 * depends on the in-context instrumented package/version.
 *
 * @param instrumentedPackage - Data about the instrumented package to use during filtering.
 * @param cwd - Current working directory. This must be inside of the root directory. This is an
 * absolute path.
 * @param root - Root-most directory. This is an absolute path.
 * @param fileExtensions - List of file extensions to capture metrics for.
 * @param logger - Logger instance.
 * @returns A (possibly empty) array of source files.
 */
export async function findRelevantSourceFiles(
  instrumentedPackage: PackageData,
  cwd: string,
  root: string,
  fileExtensions: string[],
  logger: Logger
) {
  logger.traceEnter('', 'findRelevantSourceFiles', [instrumentedPackage, cwd, root, fileExtensions])
  const sourceFiles = await getTrackedSourceFiles(root, logger, fileExtensions)

  const dependencyTree = await getDependencyTree(cwd, root, logger)

  const filterPromises = sourceFiles.map(async (f) => {
    const prefix = await getDirectoryPrefix(path.dirname(f.fileName), logger)
    const prefixPackageData = await getPackageData(prefix, root, logger)

    let packageTrees = getPackageTrees(dependencyTree, prefixPackageData)

    let instrumentedInstallVersions: string[] | undefined = undefined
    let shortestPathLength: number | undefined = undefined
    do {
      for (const tree of packageTrees) {
        const instrumentedInstallPaths = getInstalledVersionPaths(tree, instrumentedPackage.name)
        if (instrumentedInstallPaths.length > 0) {
          const pathsLength = instrumentedInstallPaths[0]?.length ?? 0
          if (shortestPathLength === undefined || pathsLength < shortestPathLength) {
            instrumentedInstallVersions = instrumentedInstallPaths.map(
              (path) => getPropertyByPath(tree, path)['version']
            )
            shortestPathLength = pathsLength
          }
        }
      }
      // did not find, go up one level for all packages
      packageTrees = packageTrees
        .map((tree) => getTreePredecessor(dependencyTree, tree['path'] as ObjectPath))
        .filter((tree) => tree !== undefined) as DependencyTree[]
    } while (shortestPathLength === undefined && packageTrees.length > 0)

    if (instrumentedInstallVersions === undefined) {
      throw new NoInstallationFoundError(instrumentedPackage.name)
    }

    return instrumentedInstallVersions.some((version) => version === instrumentedPackage.version)
  })
  const filterData = await Promise.all(filterPromises)

  const results = sourceFiles.filter((_, index) => {
    return filterData[index]
  })

  logger.traceEnter('', 'findRelevantSourceFiles', results)
  return results
}
