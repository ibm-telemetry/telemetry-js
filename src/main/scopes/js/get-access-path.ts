/*
 * Copyright IBM Corp. 2024, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as ts from 'typescript'

import type { Logger } from '../../core/log/logger.js'
import { ComplexValue } from './complex-value.js'
import { getNodeValueHandler } from './node-value-handler-map.js'

/**
 * Constructs a JsToken object from a given Identifier type AST node.
 *
 * @param node - TS node to retrieve access path for.
 * @param sourceFile - Root AST node.
 * @param logger - A logger instance.
 * @param currAccessPath - For internal use only, tracks current constructed access path.
 * @param topLevel - For internal use only, tracks top level function call.
 * @returns Access path to node represented as string array (chunks).
 */
export default function getAccessPath(
  node: ts.Node,
  sourceFile: ts.SourceFile,
  logger: Logger,
  currAccessPath: string[] = [],
  topLevel: boolean = true
): string[] {
  switch (node.kind) {
    case ts.SyntaxKind.Identifier:
      return [...currAccessPath, (node as ts.Identifier).escapedText.toString()]
    case ts.SyntaxKind.PropertyAccessExpression:
      currAccessPath.push((node as ts.PropertyAccessExpression).name.escapedText.toString())
      break
    case ts.SyntaxKind.ElementAccessExpression: {
      const argumentExpression = (node as ts.ElementAccessExpression).argumentExpression
      const data = getNodeValueHandler(argumentExpression.kind, sourceFile, logger).getData(
        argumentExpression
      )
      if (data instanceof ComplexValue) {
        currAccessPath = [
          ...currAccessPath,
          getAccessPath(argumentExpression, sourceFile, logger, [], false).reverse().join('.')
        ]
      } else {
        currAccessPath.push(data?.toString() ?? '')
      }
      break
    }
  }

  let accessPath = currAccessPath

  if ('expression' in node) {
    accessPath = getAccessPath(
      node.expression as ts.Node,
      sourceFile,
      logger,
      currAccessPath,
      false
    )
  }

  return topLevel ? accessPath.reverse() : accessPath
}
