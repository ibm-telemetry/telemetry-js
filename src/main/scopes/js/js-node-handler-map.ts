/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as ts from 'typescript'

import { JsNodeHandlerMap } from '../js/interfaces.js'
import { ImportNodeHandler } from '../js/node-handlers/import-node-handler.js'
import { CallExpressionNodeHandler } from './node-handlers/tokens-and-functions-handlers/call-expression-node-handler.js'
import { ElementAccessExpressionNodeHandler } from './node-handlers/tokens-and-functions-handlers/element-access-expression-node-handler.js'
import { IdentifierNodeHandler } from './node-handlers/tokens-and-functions-handlers/identifier-node-handler.js'
import { PropertyAccessExpressionNodeHandler } from './node-handlers/tokens-and-functions-handlers/property-access-expression-node-handler.js'

/**
 * Maps node kinds to handlers that know how to process them to generate JsFunction and JsToken
 * metrics for the JsScope.
 */
export const jsNodeHandlerMap: JsNodeHandlerMap = {
  [ts.SyntaxKind.ImportDeclaration]: ImportNodeHandler,
  [ts.SyntaxKind.CallExpression]: CallExpressionNodeHandler,
  [ts.SyntaxKind.PropertyAccessExpression]: PropertyAccessExpressionNodeHandler,
  [ts.SyntaxKind.ElementAccessExpression]: ElementAccessExpressionNodeHandler,
  [ts.SyntaxKind.Identifier]: IdentifierNodeHandler
}
