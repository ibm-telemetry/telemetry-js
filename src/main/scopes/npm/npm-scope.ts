/*
 * Copyright IBM Corp. 2023, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { type Logger } from '../../core/log/logger.js'
import { Scope } from '../../core/scope.js'
import { findInstallingPackages } from './find-installing-packages.js'
import { getInstrumentedPackageData } from './get-instrumented-package-data.js'
import { DependencyMetric } from './metrics/dependency-metric.js'

/**
 * Scope class dedicated to data collection from an npm environment.
 */
export class NpmScope extends Scope {
  protected override logger: Logger
  public name = 'npm'

  /**
   * Constructs an NpmScope.
   *
   * @param logger - Injected logger dependency.
   */
  public constructor(logger: Logger) {
    super()
    this.logger = logger
  }

  public override async run(): Promise<void> {
    const { name: instrumentedPkgName, version: instrumentedPkgVersion } =
      await getInstrumentedPackageData()

    const installingPackages = await findInstallingPackages(
      instrumentedPkgName,
      instrumentedPkgVersion
    )

    installingPackages.forEach((installingPkg) => {
      installingPkg.dependencies.forEach((dependency) => {
        this.capture(
          new DependencyMetric({
            name: dependency.name,
            version: dependency.version,
            installerName: installingPkg.name,
            installerVersion: installingPkg.version
          })
        )
      })
    })
  }
}
