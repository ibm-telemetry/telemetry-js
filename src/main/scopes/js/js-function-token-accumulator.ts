/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { JsFunction, JsImport, JsToken } from '../js/interfaces.js'
import { JsAccumulator } from '../js/js-accumulator.js'

/**
 * Responsible for maintaining an aggregated state of imports, functions and tokens.
 */
export class JsFunctionTokenAccumulator extends JsAccumulator {
  public readonly tokens: JsToken[]
  public readonly functions: JsFunction[]
  public readonly functionImports: Map<JsFunction, JsImport>
  public readonly tokenImports: Map<JsToken, JsImport>

  constructor() {
    super()
    this.tokens = []
    this.functions = []
    this.functionImports = new Map()
    this.tokenImports = new Map()
  }
}
