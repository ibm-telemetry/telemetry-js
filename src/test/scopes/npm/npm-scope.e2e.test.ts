/*
 * Copyright IBM Corp. 2023, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'node:path'

import { type ConfigSchema } from '@ibm/telemetry-config-schema'
import { describe, expect, it } from 'vitest'

import { EmptyScopeError } from '../../../main/exceptions/empty-scope.error.js'
import { NoPackageJsonFoundError } from '../../../main/exceptions/no-package-json-found-error.js'
import { NpmScope } from '../../../main/scopes/npm/npm-scope.js'
import { Fixture } from '../../__utils/fixture.js'
import { initLogger } from '../../__utils/init-logger.js'
import { initializeOtelForTest } from '../../__utils/initialize-otel-for-test.js'

const config: ConfigSchema = {
  projectId: 'abc123',
  version: 1,
  collect: { npm: { dependencies: null } }
}

describe('class: NpmScope', () => {
  const logger = initLogger()

  describe('run', () => {
    it('correctly captures dependency data', async () => {
      const fixture = new Fixture('projects/basic-project/node_modules/instrumented')
      const scope = new NpmScope(
        fixture.path,
        path.join(fixture.path, '..', '..'),
        { collect: { npm: { dependencies: null } }, projectId: '123', version: 1 },
        logger
      )

      const metricReader = initializeOtelForTest()

      await scope.run()

      const results = await metricReader.collect()

      // Metric timestamps change every run, so change them to a fixed value when comparing against
      // snapshots
      results.resourceMetrics.scopeMetrics.forEach((scopeMetric) => {
        scopeMetric.metrics.forEach((metric) => {
          metric.dataPoints.forEach((dataPoint) => {
            ;(dataPoint as { startTime: [number, number] }).startTime = [0, 0]
            ;(dataPoint as { endTime: [number, number] }).endTime = [0, 0]
          })
        })
      })

      expect(results).toMatchSnapshot()
    })

    it('throws EmptyScopeError if no collector has been defined', async () => {
      const fixture = new Fixture('projects/basic-project/node_modules/instrumented')
      const scope = new NpmScope(
        fixture.path,
        path.join(fixture.path, '..', '..'),
        { collect: { npm: {} }, projectId: '123', version: 1 },
        logger
      )

      await expect(scope.run()).rejects.toThrow(EmptyScopeError)
    })
  })

  it('throws an error if no package.json files were found', async () => {
    const fixture = new Fixture('projects/no-package-json-files/foo/bar')
    const scope = new NpmScope(
      fixture.path,
      path.join(fixture.path, '..', '..'),
      { collect: { npm: {} }, projectId: '123', version: 1 },
      logger
    )

    await expect(
      async () => await scope.findInstallingPackages('not-here', '0.1.0')
    ).rejects.toThrow(NoPackageJsonFoundError)
  })

  describe('findInstallingPackages', () => {
    it('correctly finds installing package data', async () => {
      const fixture = new Fixture('projects/basic-project/node_modules/instrumented')
      const pkgs = await new NpmScope(
        fixture.path,
        path.join(fixture.path, '..', '..'),
        config,
        logger
      ).findInstallingPackages('instrumented', '0.1.0')

      expect(pkgs).toMatchSnapshot()
    })

    it('finds no results for an unknown package', async () => {
      const fixture = new Fixture('projects/basic-project/node_modules/instrumented')
      const pkgs = await new NpmScope(
        fixture.path,
        path.join(fixture.path, '..', '..'),
        config,
        logger
      ).findInstallingPackages('not-here', '0.1.0')

      expect(pkgs).toMatchSnapshot()
    })

    it('finds no results for an known package at an unknown version', async () => {
      const fixture = new Fixture('projects/basic-project/node_modules/instrumented')
      const pkgs = await new NpmScope(
        fixture.path,
        path.join(fixture.path, '..', '..'),
        config,
        logger
      ).findInstallingPackages('instrumented', '0.3.0')

      expect(pkgs).toMatchSnapshot()
    })
  })
})
