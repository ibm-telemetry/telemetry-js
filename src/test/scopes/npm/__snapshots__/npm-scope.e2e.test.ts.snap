// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`class: NpmScope > findInstallingPackages > correctly finds installing package data 1`] = `
[
  {
    "dependencies": [
      {
        "name": "foo",
        "version": "1.0.0",
      },
      {
        "name": "instrumented",
        "version": "0.1.0",
      },
    ],
    "name": "basic-project",
    "version": "1.0.0",
  },
]
`;

exports[`class: NpmScope > findInstallingPackages > finds no results for an known package at an unknown version 1`] = `[]`;

exports[`class: NpmScope > findInstallingPackages > finds no results for an unknown package 1`] = `[]`;

exports[`class: NpmScope > run > correctly captures dependency data 1`] = `
{
  "errors": [],
  "resourceMetrics": {
    "resource": Resource {
      "_asyncAttributesPromise": undefined,
      "_attributes": {
        "analyzed.host": "host",
        "analyzed.owner": "owner",
        "analyzed.raw": "gitOrigin",
        "analyzed.repository": "repository",
        "date": "1970-01-01T00:00:02.023Z",
        "project.id": "projectId",
        "service.name": "IBM Telemetry",
        "telemetry.emitter.name": "telemetryName",
        "telemetry.emitter.version": "telemetryVersion",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "[omitted for test snapshot]",
      },
      "_syncAttributes": {
        "analyzed.host": "host",
        "analyzed.owner": "owner",
        "analyzed.raw": "gitOrigin",
        "analyzed.repository": "repository",
        "date": "1970-01-01T00:00:02.023Z",
        "project.id": "projectId",
        "service.name": "IBM Telemetry",
        "telemetry.emitter.name": "telemetryName",
        "telemetry.emitter.version": "telemetryVersion",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "[omitted for test snapshot]",
      },
      "asyncAttributesPending": false,
    },
    "scopeMetrics": [
      {
        "metrics": [
          {
            "aggregationTemporality": 1,
            "dataPointType": 3,
            "dataPoints": [
              {
                "attributes": {
                  "npm.dependency.installer.name": "e497ed4e62a3b47d9cd8aaebefb6c0837b94fb8d710461ba203aa33ccf64cc4d",
                  "npm.dependency.installer.raw": "e497ed4e62a3b47d9cd8aaebefb6c0837b94fb8d710461ba203aa33ccf64cc4d",
                  "npm.dependency.installer.version.major": "1",
                  "npm.dependency.installer.version.minor": "0",
                  "npm.dependency.installer.version.patch": "0",
                  "npm.dependency.installer.version.raw": "92521fc3cbd964bdc9f584a991b89fddaa5754ed1cc96d6d42445338669c1305",
                  "npm.dependency.name": "2c26b46b68ffc68ff99b453c1d30413413422d706483bfa0f98a5e886266e7ae",
                  "npm.dependency.raw": "2c26b46b68ffc68ff99b453c1d30413413422d706483bfa0f98a5e886266e7ae",
                  "npm.dependency.version.major": "1",
                  "npm.dependency.version.minor": "0",
                  "npm.dependency.version.patch": "0",
                  "npm.dependency.version.raw": "92521fc3cbd964bdc9f584a991b89fddaa5754ed1cc96d6d42445338669c1305",
                },
                "endTime": [
                  0,
                  0,
                ],
                "startTime": [
                  0,
                  0,
                ],
                "value": 1,
              },
              {
                "attributes": {
                  "npm.dependency.installer.name": "e497ed4e62a3b47d9cd8aaebefb6c0837b94fb8d710461ba203aa33ccf64cc4d",
                  "npm.dependency.installer.raw": "e497ed4e62a3b47d9cd8aaebefb6c0837b94fb8d710461ba203aa33ccf64cc4d",
                  "npm.dependency.installer.version.major": "1",
                  "npm.dependency.installer.version.minor": "0",
                  "npm.dependency.installer.version.patch": "0",
                  "npm.dependency.installer.version.raw": "92521fc3cbd964bdc9f584a991b89fddaa5754ed1cc96d6d42445338669c1305",
                  "npm.dependency.name": "ec46e364d52dab2e207d33617267a84df5838926793c1b0a2974899fe28229f1",
                  "npm.dependency.raw": "ec46e364d52dab2e207d33617267a84df5838926793c1b0a2974899fe28229f1",
                  "npm.dependency.version.major": "0",
                  "npm.dependency.version.minor": "1",
                  "npm.dependency.version.patch": "0",
                  "npm.dependency.version.raw": "6ad9613a455798d6d92e5f5f390ab4baa70596bc869ed6b17f5cdd2b28635f06",
                },
                "endTime": [
                  0,
                  0,
                ],
                "startTime": [
                  0,
                  0,
                ],
                "value": 1,
              },
            ],
            "descriptor": {
              "advice": {},
              "description": "",
              "name": "dependency.count",
              "type": "COUNTER",
              "unit": "",
              "valueType": 0,
            },
            "isMonotonic": true,
          },
        ],
        "scope": {
          "name": "npm",
          "schemaUrl": undefined,
          "version": "",
        },
      },
    ],
  },
}
`;
